// @electron-forge/web-multi-logger/dist/Logger.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "Tab", {
  enumerable: true,
  get: function () {
    return _Tab.default;
  }
});
exports.default = void 0;

require("source-map-support/register");

var _express = _interopRequireDefault(require("express"));

var _path = _interopRequireDefault(require("path"));

var _expressWs = _interopRequireDefault(require("express-ws"));

var _Tab = _interopRequireDefault(require("./Tab"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class Logger {
  constructor(port = 9000) {
    this.port = port;

    _defineProperty(this, "app", (0, _express.default)());

    _defineProperty(this, "ws", void 0);

    _defineProperty(this, "tabs", []);

    _defineProperty(this, "server", null);

    this.registerRoutes();
  }

  registerRoutes() {
    this.ws = (0, _expressWs.default)(this.app);
    this.app.get('/rest/tabs', (req, res) => res.json(this.tabs));
    this.app.use('/xterm', _express.default.static(_path.default.resolve(require.resolve('xterm'), '../../../dist')));
    this.app.use(_express.default.static(_path.default.resolve(__dirname, '..', 'static')));
    this.app.ws('/sub', () => {});
  }
  /**
   * Creates a new tab with the given name, the name should be human readable
   * it will be used as the tab title in the front end.
   */


  createTab(name) {
    const tab = new _Tab.default(name, this.ws);
    this.tabs.push(tab);
    return tab;
  }
  /**
   * Start the HTTP server hosting the web UI
   */


  start() {
    return new Promise(resolve => {
      this.server = this.app.listen(this.port, () => resolve(this.port));
    });
  }
  /**
   * Stop the HTTP server hosting the web UI
   */


  stop() {
    if (this.server) this.server.close();
  }

}

exports.default = Logger;
