// @electron-forge/web-multi-logger/dist/Tab.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

require("source-map-support/register");

var _Log = _interopRequireDefault(require("./Log"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

let idCounter = 1;

class Tab {
  constructor(name, ws) {
    this.name = name;
    this.ws = ws;

    _defineProperty(this, "logs", []);

    _defineProperty(this, "id", void 0);

    this.id = idCounter;
    idCounter += 1;
  }
  /**
   * Log a line to the web UI, a new line is automatically appended to the line
   */


  log(line) {
    const log = new _Log.default(line, new Date());
    this.logs.push(log);

    for (const client of this.ws.getWss().clients) {
      client.send(JSON.stringify({
        tab: this.id,
        payload: log
      }));
    }
  }

  toJSON() {
    return {
      id: this.id,
      name: this.name,
      logs: this.logs
    };
  }

}

exports.default = Tab;
